### PR TITLE
remove read lock in meta client

### LINKED ISSUE #3129: Remove all read-write locks in meta client
from @critical27

### PR DESCRIPTION
**Pull request by jackwener**
#### What type of PR is this?
- [x] feature

#### Which issue(s) this PR fixes:
close #3129 

#### What this PR does / why we need it?
reduce the side effect of  ` read-write locks `

#### Checklist：
- [x] Performance regression: Consumes more Memory                                           `

### CODE REVIEW

**Review by critical27 — COMMENTED** [review]
Good job. After I review the code, some places you need to pay attention:
1. You only handle the structure in `SpaceInfoCache`, I think some function is missing, for example `getTagSchemaFromCache`, `getLatestTagVersionFromCache` and so on. In fact, almost all function with `folly::RWSpinLock::ReadHolder` should be taken care.
2. Just delete the commented code, don't keep them in the old place

You could do some simple performance test using k6 or something else. This PR is very sensitive to performance.

**Comment by jackwener** [comment]
There are 5 remain read lock in 667 2319 2360 2383 2395 line.
`leader` is in line 667. `leader` is updated in `updateLeader()`  outside `loaddata()`.
`leadersInfo_` is in line 2319 and 2360. `leadersInfo_.leaderMap_` and `leadersInfo_.pickedIndex_` is update outside `loaddata()`.
`authCheckFromCache()` is in 2383 and `checkShadowAccountFromCache()` is in 2395, they must be `const` because other const function use it

File: src/clients/meta/MetaClient.cpp [hunk 1]
```diff
@@ -438,11 +476,22 @@ bool MetaClient::loadSchemas(GraphSpaceID spaceId,
...
   }
 
-  spaceInfoCache->tagSchemas_ = std::move(tagSchemas);
-  spaceInfoCache->edgeSchemas_ = std::move(edgeSchemas);
+  // spaceInfoCache->tagSchemas_ = std::move(tagSchemas);
+  // spaceInfoCache->edgeSchemas_ = std::move(edgeSchemas);
   return true;
 }
 
+static Indexes __buildIndexes(std::vector<cpp2::IndexItem> indexItemVec) {
```
Comment (CPWstatic) [inline src/clients/meta/MetaClient.cpp:484]: Please pay attention to the name style, we don't use `__` as prefix.

File: src/clients/meta/MetaClient.cpp [hunk 2]
```diff
@@ -399,21 +437,21 @@ bool MetaClient::loadSchemas(GraphSpaceID spaceId,
...
+    //   addSchemaField(schema.get(), col, &spaceInfoCache->pool_);
+    // }
+    // // handle shcem property
+    // schema->setProp(edgeIt.get_schema().get_schema_prop());
+    // if (edgeIt.get_edge_type() != lastEdgeType) {
+    //   // init schema vector, since schema version is zero-based, need to add one
+    //   edgeSchemas[edgeIt.get_edge_type()].resize(schema->getVersion() + 1);
+    //   lastEdgeType = edgeIt.get_edge_type();
+    // }
+    // edgeSchemas[edgeIt.get_edge_type()][schema->getVersion()] = std::move(schema);
```
Comment (CPWstatic) [inline src/clients/meta/MetaClient.cpp:454]: Just delete them, if you are sure that you are doing the right thing.

**Review by CPWstatic — COMMENTED** [review]
Good job!

**Comment by critical27** [comment]
> There are 5 remain read lock in 667 2319 2360 2383 2395 line. `leader` is in line 667. `leader` is updated in `updateLeader()` outside `loaddata()`. `leadersInfo_` is in line 2319 and 2360. `leadersInfo_.leaderMap_` and `leadersInfo_.pickedIndex_` is update outside `loaddata()`. `authCheckFromCache()` is in 2383 and `checkShadowAccountFromCache()` is in 2395, they must be `const` because const client in `PasswordAuthenticator()` call it.

I see, my bad, Github has fold some function name...

**Review by critical27 — APPROVED** [review]
Well done, LGTM

**Review by CPWstatic — APPROVED** [review]

**Pull request merged**